tls/s2n_connection.c

```diff
@@ -1161,13 +1161,17 @@ const char *s2n_get_application_protocol(struct s2n_connection *conn)
 int s2n_connection_get_session_id_length(struct s2n_connection *conn)
 {
     POSIX_ENSURE_REF(conn);
+    /* Stateful session resumption in TLS1.3 using session id is not yet supported. */
+    POSIX_ENSURE(conn->actual_protocol_version < S2N_TLS13, S2N_ERR_UNIMPLEMENTED);
     return conn->session_id_len;
 }
 
 int s2n_connection_get_session_id(struct s2n_connection *conn, uint8_t *session_id, size_t max_length)
 {
     POSIX_ENSURE_REF(conn);
     POSIX_ENSURE_REF(session_id);
+    /* Stateful session resumption in TLS1.3 using session id is not yet supported. */
+    POSIX_ENSURE(conn->actual_protocol_version < S2N_TLS13, S2N_ERR_UNIMPLEMENTED);
 
```

Comment:
Do we fail this call in TLS1.2 if the session cache is not in use?

Also, this is a behavior change. Will it break customers who aren't expecting s2n_connection_get_session_id to start failing?

Reply:
We don't fail in TLS1.2 as session resumption with stateful is possible. In the case of TLS1.3 the data returned cannot be used meaningfully. The API can behave the same way, but the value returned is not usable. 
 For these two APIs, we can possibly ignore and keep the behavior but the APIs mentioned here https://github.com/aws/s2n-tls/issues/2553#issuecomment-840815356  will fail for stateful session resumption in TLS1.3 and behavior cannot be the same as TLS1.2 until we implement the stateful TLS1.3 session resumption feature.

Reply:
Possible, but not always enabled. Why is TLS1.2 without stateful resumption enabled any less of an error than TLS1.3?